agent/src/util/registration.rs

```diff
@@ -0,0 +1,458 @@
+use ::url::Url;
+use akri_discovery_utils::discovery::{
+    v0::{
+        registration_server::{Registration, RegistrationServer},
+        Empty, RegisterRequest,
+    },
+    AGENT_REGISTRATION_SOCKET, DISCOVERY_HANDLER_PATH,
+};
+use akri_shared::uds::unix_stream;
+use futures::TryStreamExt;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use std::time::Instant;
+use tokio::sync::broadcast;
+use tonic::{transport::Server, Request, Response, Status};
+
+/// Maximum amount of time allowed to pass without being able to connect to a discovery handler
+/// without it being removed from the map of registered Discovery Handlers.
+pub const DISCOVERY_HANDLER_OFFLINE_GRACE_PERIOD_SECS: u64 = 300;
+
+/// Fake endpoint that signals a DiscoveryOperator to use an embedded
+/// discovery handler.
+pub const EMBEDDED_DISCOVERY_HANDLER_ENDPOINT: &str = "embedded";
+
+// Map of RegisterRequests for a specific protocol where key is the endpoint of the Discovery Handler
+// and value is whether or not the discovered devices are local.
+pub type ProtocolDiscoveryHandlerMap = HashMap<DiscoveryHandlerEndpoint, DiscoveryHandlerDetails>;
+
+/// Map of all registered Discovery Handlers where key is protocol
+/// and value is a map of all Discovery Handlers for that protocol.
+pub type RegisteredDiscoveryHandlerMap = Arc<Mutex<HashMap<String, ProtocolDiscoveryHandlerMap>>>;
+
+/// A Discovery Handler's endpoint, distinguished by URI type
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum DiscoveryHandlerEndpoint {
+    /// Embedded means the Discovery Handler is running inside the Agent
+    Embedded,
+    /// Uds means the Discovery Handler is running on a specified unix domain socket
+    Uds(String),
+    /// Network means the Discovery Handler is running at an specified URL
+    Network(String),
+}
+
+/// Describes the connectivity status of a Discovery Handler.
+#[derive(PartialEq, Debug, Clone)]
+pub enum DiscoveryHandlerConnectivityStatus {
+    /// Has a client successfully using it
+    HasClient,
+    /// Registered but does not have a client
+    Online,
+    /// Not returning discovery results
+    Offline(Instant),
+}
+
+#[derive(Debug, Clone)]
+pub struct DiscoveryHandlerDetails {
+    pub register_request: RegisterRequest,
+    pub stop_discovery: broadcast::Sender<()>,
+    pub connectivity_status: DiscoveryHandlerConnectivityStatus,
+}
+
+pub fn get_external_endpoint_type(
+    endpoint: &str,
+) -> Result<DiscoveryHandlerEndpoint, url::ParseError> {
+    if endpoint.starts_with(DISCOVERY_HANDLER_PATH) {
```

Comment:
I think it makes sense to put the burden on the Agent rather than Discovery Handler implementers. Having a simpler gRPC interface seems good, but being explicit could catch errors. I also anticipate 99.9% of discovery handlers using UDS.

Reply:
i'd rather be explicit than have the agent parse a string

Reply:
fixed in a2b2b8c